Feature/gino

thomas-maschler opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help -->

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [x] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [x] Check the commit's or even all commits' message styles matches our requested structure.
- [x] Check your code additions will fail neither code linting checks nor unit test.



## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number:
- [GTC-599]
- [GTC-600]


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

1. Update tile cache
- Remove feature, max_date and geostore routes
- Use Gino as main async db framework to fetch vector tiles
- Make direct calls to API database to fetch version and asset metadata using synchronous calls
- Refactor code base to match data-api structure
- Add hidden helper endpoint to fetch latest versions
- Increase test coverage
- Move redoc to root
- Add CORS and GZIP middleware
- Adopt JSEND schema for JSON response (WIP, still need to work on error messages [GTC-628], ESRI Vector Tile Service are exempt)

2. Update terraform
- Disable workspaces for feature branches (one default workspace in DEV!)
- Update cloud front routing behavior
- Use external module for fargate 
- Update Lambda@Edge functions to redirect `latest` version to correct version number
- Manage log groups for lambda@Edge functions
- Add ACM certificate in staging

3. CI/CD
- Use Pipenv to manage dependencies
- Update ci/cd scripts
- Add Github Actions Workflows

## Does this introduce a breaking change?

- [x] Yes
- [] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->
Remove feature, max_date and geostore routes. They now live with DATA-API

## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

There are still some issues which I didn't address in the PR. 
Routes for custom services such as fires are hard coded. B/C I have to look up fields and available versions upfront, the service will crash if no such dataset is in the database. Hence before we merge into develop/ master, we will need to create the viirs dataset via the Data API in the different environments.
I created a separate ticket to address this issue [GTC-627] (not just for fires but any dynamic tile cache with custom filters). But this is a larger project.

jterry64 commented on app/crud/async_db/vector_tiles/filters.py:
```diff
@@ -0,0 +1,77 @@
+
+
+async def geometry_filter(
+    geostore_id: Optional[str], bounds: Bounds
```
Why is geostore_id optional? Is it supposed to do something if it's None (like just return the whole bounding box)?

thomas-maschler replied:
All the filters (except date) are optional.
In the route I simply forward the user input (which is either `None` or a value) to the filter. Only here i check if the value is empty and if yes return and and convert it into a Text Clause. Only here I check the value and return `None` or a TextClause.

In the route I then only filter out empty filters.

https://github.com/wri/gfw-tile-cache/blob/feature/gino/app/routes/nasa_viirs_fire_alerts/vector_tiles.py#L79-L86

Filter out `None` values in the route before converting them to TextClause seem to verbose to me.

jterry64 commented on app/crud/async_db/vector_tiles/filters.py:
```diff
@@ -0,0 +1,77 @@
+
+
+def date_filter(start_date: str, end_date: str) -> TextClause:
+    f: TextClause = db.text(
```
should this accept the date field name too, since "alert__date" might not generically be the date field for other vector tables?

thomas-maschler replied:
right

jterry64 commented on app/crud/async_db/vector_tiles/filters.py:
```diff
@@ -0,0 +1,77 @@
+
+
+def confidence_filter(high_confidence_only: Optional[bool]) -> Optional[TextClause]:
+    if high_confidence_only:
```
I would move this to nasa_viirs_fire_alerts.py since it's only related to VIIRS

thomas-maschler replied:
i guess we can even replace it with a generic contextual filter once we made sure that fire data are normalized

jterry64 commented on app/routes/__init__.py:
```diff
@@ -0,0 +1,150 @@
+DEFAULT_END = NOW.to_date_string()
+
+Geometry = Dict[str, Any]
+Bounds = Tuple[float, float, float, float]
```
You define these types in a couple places, is it possible to just define them from one place?

thomas-maschler commented:
ok, addressed your comments @jterry64. I will still wait before deploying to staging b/c it currently requires present of the viirs dataset in the database.

We will first need to load the data into data api before I can deploy.

thomas-maschler commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/crud/sync_db/versions.py  4
- app/crud/async_db/vector_tiles/__init__.py  2
- app/routes/helpers.py  1
- app/models/pydantic/nasa_viirs_fire_alerts.py  1
- app/models/enumerators/dynamic_enumerators.py  3
- app/routes/__init__.py  5
- app/routes/nasa_viirs_fire_alerts/vector_tiles.py  3
- app/crud/sync_db/vector_tile_assets.py  4
- terraform/modules/content_delivery_network/lambda_functions/redirect_latest_tile_cache/src/lambda_function.py  4
- app/routes/nasa_viirs_fire_alerts/__init__.py  3
- app/main.py  1
- app/routes/dynamic_vector_tiles.py  2
- app/routes/raster_tiles.py  1
- app/models/pydantic/versions.py  1
- app/utils/geostore.py  3
- app/crud/async_db/vector_tiles/nasa_viirs_fire_alerts.py  2
- app/routes/vector_tiles.py  1
- app/models/pydantic/esri.py  1
- tests/conftest.py  3
- app/routes/nasa_viirs_fire_alerts/esri_vector_tile_server.py  3
- app/models/pydantic/responses.py  1
- app/models/pydantic/database.py  4
- terraform/modules/content_delivery_network/lambda_functions/redirect_s3_404/src/lambda_function.py  4
- app/responses.py  1
- app/routes/esri_vector_tile_server.py  4
- app/crud/async_db/vector_tiles/filters.py  3
- app/application.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wri/gfw-tile-cache/pullRequest?prid=5685621&bid=18609033)

thomas-maschler commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/routes/nasa_viirs_fire_alerts/__init__.py  3
- terraform/modules/content_delivery_network/lambda_functions/redirect_latest_tile_cache/src/lambda_function.py  4
- app/responses.py  1
- terraform/modules/content_delivery_network/lambda_functions/redirect_s3_404/src/lambda_function.py  4
- app/routes/dynamic_vector_tiles.py  2
- app/crud/async_db/vector_tiles/filters.py  3
- app/models/enumerators/dynamic_enumerators.py  3
- app/crud/async_db/vector_tiles/nasa_viirs_fire_alerts.py  2
- app/routes/helpers.py  1
- app/utils/geostore.py  3
- app/models/pydantic/nasa_viirs_fire_alerts.py  1
- app/routes/nasa_viirs_fire_alerts/vector_tiles.py  3
- app/crud/async_db/vector_tiles/__init__.py  2
- app/routes/esri_vector_tile_server.py  4
- app/routes/nasa_viirs_fire_alerts/esri_vector_tile_server.py  3
- app/routes/raster_tiles.py  1
- app/main.py  1
- tests/conftest.py  3
- app/routes/__init__.py  5
- app/models/pydantic/versions.py  1
- app/routes/vector_tiles.py  1
- app/crud/sync_db/versions.py  4
- app/crud/sync_db/vector_tile_assets.py  4
- app/models/pydantic/responses.py  1
- app/models/pydantic/database.py  4
- app/application.py  3
- app/models/pydantic/esri.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wri/gfw-tile-cache/pullRequest?prid=5685621&bid=18609033)
